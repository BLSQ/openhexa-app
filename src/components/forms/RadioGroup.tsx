import clsx from "clsx";
import { FormEvent, HTMLAttributes } from "react";

interface RadioOption {
  id: OptionId;
  label: string;
}

type OptionId = string;

interface RadioGroupProps extends HTMLAttributes<HTMLInputElement> {
  options: RadioOption[];
  value?: OptionId;
  name: string;
}

const RadioGroup = ({ name, options, onChange, value }: RadioGroupProps) => {
  return (
    <fieldset className="flex gap-4">
      {options.map((option) => (
        <div key={option.id} className="flex items-center">
          <input
            id={`${name}-${option.id}`}
            type="radio"
            name={name}
            className={clsx(
              "form-radio focus:ring-transparent focus:outline-none text-gray-800"
            )}
            value={option.id}
            onChange={onChange}
            defaultChecked={option.id === value}
          />
          <label
            htmlFor={`${name}-${option.id}`}
            className="text-sm font-medium ml-2 text-gray-700"
          >
            {option.label}
          </label>
        </div>
      ))}
    </fieldset>
  );
};

export default RadioGroup;
