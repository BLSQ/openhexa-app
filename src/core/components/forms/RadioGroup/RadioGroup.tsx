import clsx from "clsx";
import { FormEvent, HTMLAttributes } from "react";

interface RadioOption {
  id: OptionId;
  label: string;
}

type OptionId = string;

interface RadioGroupProps extends HTMLAttributes<HTMLInputElement> {
  options: RadioOption[];
  value?: OptionId;
  disabled?: boolean;
  name: string;
}

const RadioGroup = ({
  name,
  options,
  disabled = false,
  onChange,
  value,
}: RadioGroupProps) => {
  return (
    <fieldset className="flex gap-4">
      {options.map((option) => (
        <div key={option.id} className="flex items-center">
          <input
            id={`${name}-${option.id}`}
            type="radio"
            name={name}
            disabled={disabled}
            className={clsx(
              "form-radio text-gray-800 focus:outline-none focus:ring-transparent"
            )}
            value={option.id}
            onChange={onChange}
            defaultChecked={option.id === value}
          />
          <label
            htmlFor={`${name}-${option.id}`}
            className="ml-2 text-sm font-medium text-gray-700"
          >
            {option.label}
          </label>
        </div>
      ))}
    </fieldset>
  );
};

export default RadioGroup;
