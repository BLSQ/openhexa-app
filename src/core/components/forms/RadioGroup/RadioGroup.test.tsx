import RadioGroup from "./RadioGroup";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const options = [
  { id: "1", label: "Option 1" },
  { id: "2", label: "Option 2" },
  { id: "3", label: "Option 3" },
];

describe("RadioGroup", () => {
  it("renders", async () => {
    const onChange = jest.fn();
    const { container, debug } = render(
      <RadioGroup
        name="radio"
        onChange={onChange}
        value={"2"}
        options={options}
      />
    );

    const selectedOption: HTMLInputElement | null =
      container.querySelector('input[checked=""]');

    expect(selectedOption).not.toBeNull();
    expect(selectedOption!.getAttribute("value")).toEqual("2");
    expect(container).toMatchSnapshot();

    await userEvent.click(screen.getByLabelText("Option 1"));

    expect(onChange).toHaveBeenCalled();
    expect(onChange.mock.calls[0][0].target.value).toBe("1");
  });

  it("does not trigger onChange if disabled", async () => {
    const onChange = jest.fn();
    const { container, debug } = render(
      <RadioGroup
        name="radio"
        onChange={onChange}
        disabled
        value={"2"}
        options={options}
      />
    );

    await userEvent.click(screen.getByLabelText("Option 1"));

    expect(onChange).not.toHaveBeenCalled();
  });
});
