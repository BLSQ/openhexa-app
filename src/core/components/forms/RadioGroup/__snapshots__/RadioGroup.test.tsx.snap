// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup renders 1`] = `
<div>
  <fieldset
    class="flex gap-4"
  >
    <div
      class="flex items-center"
    >
      <input
        class="form-radio text-gray-800 focus:outline-none focus:ring-transparent"
        id="radio-1"
        name="radio"
        type="radio"
        value="1"
      />
      <label
        class="ml-2 text-sm font-medium text-gray-700"
        for="radio-1"
      >
        Option 1
      </label>
    </div>
    <div
      class="flex items-center"
    >
      <input
        checked=""
        class="form-radio text-gray-800 focus:outline-none focus:ring-transparent"
        id="radio-2"
        name="radio"
        type="radio"
        value="2"
      />
      <label
        class="ml-2 text-sm font-medium text-gray-700"
        for="radio-2"
      >
        Option 2
      </label>
    </div>
    <div
      class="flex items-center"
    >
      <input
        class="form-radio text-gray-800 focus:outline-none focus:ring-transparent"
        id="radio-3"
        name="radio"
        type="radio"
        value="3"
      />
      <label
        class="ml-2 text-sm font-medium text-gray-700"
        for="radio-3"
      >
        Option 3
      </label>
    </div>
  </fieldset>
</div>
`;
